hazelcast/src/main/java/com/hazelcast/sql/impl/type/converter/DoubleConverter.java

```diff
@@ -38,27 +40,65 @@ private DoubleConverter() {
 
     @Override
     public byte asTinyint(Object val) {
-        return (byte) cast(val);
+        double casted = cast(val);
+        // here the overflow may happen: (byte) casted = (byte) (int) casted
+        byte converted = (byte) casted;
+
+        // casts from double to int are saturating
+        if (converted != (int) casted || !Double.isFinite(casted)) {
+            throw numericOverflow(QueryDataTypeFamily.TINYINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public short asSmallint(Object val) {
-        return (short) cast(val);
+        double casted = cast(val);
+        // here the overflow may happen: (short) casted = (short) (int) casted
+        short converted = (short) casted;
+
+        // casts from double to int are saturating
+        if (converted != (int) casted || !Double.isFinite(casted)) {
+            throw numericOverflow(QueryDataTypeFamily.SMALLINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public int asInt(Object val) {
-        return (int) cast(val);
+        double casted = cast(val);
+        int converted = (int) casted;
+
+        // casts from double to long are saturating
+        if (converted != (long) casted || !Double.isFinite(casted)) {
+            throw numericOverflow(QueryDataTypeFamily.INT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public long asBigint(Object val) {
-        return (long) cast(val);
+        double casted = cast(val);
+        double truncated = casted > 0.0 ? Math.floor(casted) : Math.ceil(casted);
+        // casts from double to long are saturating
+        long converted = (long) truncated;
+
+        // No checks for NaNs and infinities are needed: NaNs are zeros and
+        // infinities are Long.MAX/MIN_VALUE when converted to long.
+        if ((double) converted != truncated) {
+            throw numericOverflow(QueryDataTypeFamily.BIGINT, val);
```

Comment:
We throw overflow exceptions for NaN/Infinite. However, it seems that conversion to VARCHAR works fine in this case. Consider the following statement
```
SELECT ... CAST((CAST doubleCol as VARCHAR) as DOUBLE) ...
```
What would happen if the value of the column is `NaN`? Will it convert the `NaN` string back to `Double.NaN`?

Reply:
Yes, it will convert `Double.NaN` to `'NaN'` and then back to `Double.NaN`, other special values are working the same way. Why you are considering it as an issue? Double-to-string is lossless and reversible, but double-to-bigint is lossy and not reversible because of that.

Reply:
My concern was it is possible to convert "NaN" from VARCHAR back to REAL/DOUBLE. If this is the case, then I do not see any issues.